### DISCUSSION TITLE
Is it possible to return Error Response which Content-Type is text/html?

### DISCUSSION THREAD

**Discussion by kitagry** [goadesign/goa#2860] [Q&A] [answered]
Hi maintainers.

I want to return the error response which Content-Type is text/html. I try the followings, but this returns errors.
I think this error happen in [here](
https://github.com/goadesign/goa/blob/20b02e80edc370f29814f129966e0015800146fa/http/encoding.go#L268-L280). And this is because we can't return error which type is `[]byte`, `string` or `*string`. Is there any good way?

Thanks!

code:

```go
// design
Error("InternalServerError", String)

HTTP(func() {
  GET("test")
  Response("InternalServerError", StatusInternalServerError, func() { ContentType("text/html") })
})

// service
return nil, test.InternalServerError("<p>internal server error</p>")
```

error:

```
encoding: can't encode *server.TestInternalServerErrorResponseBody as text/html
```

**Comment by raphael** [comment, marked as answer]
Hello, the `text/html` encoder only supports strings or slice of bytes, it doesn't know how to serialize an arbitrary object into HTML (there really isn't a well defined way to do that in general). You can either change the error type to be a string or provide your own encoder that knows how to serialize the error object into HTML.

**Reply by kitagry** [reply]
Thank you for the response!!
I understand how to provide my own encoder.

<details><summary>Maybe like followings</summary>

```go
// design
Error("CustomError", string)
HTTP(func() {
  Response("CustomError", StatusInternalServerError, func() { ContentType("customContentType") })
})

// http.go
enc = func(ctx context.Context, w http.ResponseWriter) goahttp.Encoder {
	ct, ok := ctx.Value(goahttp.ContentTypeKey).(string)
	if !ok {
		return goahttp.ResponseEncoder(ctx, w)
	}
                                                                         
	if ct == "customContentType" {
		return NewCustomEncoder(ctx, w)
	}
	return goahttp.ResponseEncoder(ctx, w)
}
```

</details>

But, I don't understand the way to change the error type to be a string. Would you teach me?

Thank you!

**Reply by raphael** [reply]
Something like this should work:

```go
package design

import . "goa.design/goa/v3/dsl"

var _ = Service("service", func() {
    Method("method", func() {
        Error("CustomError", String)
        HTTP(func() {
            GET("/")
            Response("CustomError", StatusBadRequest)
        })
    })
})
```

Errors can then be returned like so:

```go
func (s *servicesrvc) Method(ctx context.Context) (err error) {
        return service.CustomError("doh!")
}
```

**Reply by kitagry** [reply]
Thank you for the comment!
This returns text error message, but it is handled as `application/json`.

```
$ curl http://localhost:8080/test -i
HTTP/1.1 500 Internal Server Error
Content-Type: application/json
Goa-Error: InternalError
Date: Fri, 04 Jun 2021 00:33:17 GMT
Content-Length: 7

"doh!"
```

**Reply by raphael** [reply]
It should be possible to override the content type in the response like you had above (or using a custom encoder).

**Reply by kitagry** [reply]
I'm sorry for the late reply. I understand it! Thank you!
